drivers/adc/adc_tla2021.c

```diff
@@ -0,0 +1,272 @@
+/*
+ * Copyright (c) 2023 Caspar Friedrich <c.s.w.friedrich@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/adc.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/sys/util_macro.h>
+
+#define DT_DRV_COMPAT ti_tla2021
+
+LOG_MODULE_REGISTER(tla2021, CONFIG_ADC_LOG_LEVEL);
+
+/*
+ * Conversion Data Register (RP = 00h) [reset = 0000h]
+ */
+#define REG_DATA     0x00
+#define REG_DATA_pos 4
+
+/*
+ * Configuration Register (RP = 01h) [reset = 8583h]
+ */
+#define REG_CONFIG          0x01
+#define REG_CONFIG_DEFAULT  0x8583
+#define REG_CONFIG_DR_pos   5
+#define REG_CONFIG_DR_msk   (BIT_MASK(3) << REG_CONFIG_DR_pos)
+#define REG_CONFIG_MODE_pos 8
+#define REG_CONFIG_PGA_pos  9  /* TLA2022 and TLA2024 Only */
+#define REG_CONFIG_MUX_pos  12 /* TLA2024 Only */
+#define REG_CONFIG_OS_pos   15
+#define REG_CONFIG_OS_msk   (BIT_MASK(1) << REG_CONFIG_OS_pos)
+
+#define EVENT_START_CONVERSION BIT(0)
+
+struct tla2021_config {
+	const struct i2c_dt_spec i2c_spec;
+};
+
+struct tla2021_data {
+	struct k_event event;
+	struct k_timer timer;
+	struct k_mutex lock;
+	uint16_t reg_config;
+};
+
+static int read(const struct device *dev, uint8_t reg, uint16_t *data)
+{
+	int ret;
+
+	const struct tla2021_config *config = dev->config;
+	uint8_t data_[2];
+
+	ret = i2c_write_read_dt(&config->i2c_spec, &reg, sizeof(reg), data_, sizeof(data_));
+	if (ret) {
+		LOG_ERR("Call `i2c_write_read_dt` failed: %d", ret);
+		return ret;
+	}
+
+	*data = sys_get_be16(data_);
+
+	return 0;
+}
+
+static int write(const struct device *dev, uint8_t reg, uint16_t data)
+{
+	int ret;
+
+	const struct tla2021_config *config = dev->config;
+	uint8_t data_[3] = {reg};
+
+	sys_put_be16(data, &data_[1]);
+
+	ret = i2c_write_dt(&config->i2c_spec, data_, sizeof(data_));
+	if (ret) {
+		LOG_ERR("Call `i2c_write_dt` failed: %d", ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+int tla2021_channel_setup(const struct device *dev, const struct adc_channel_cfg *cfg)
+{
+	struct tla2021_data *data = dev->data;
+
+	k_mutex_lock(&data->lock, K_FOREVER);
```

Comment:
Might be better to use a spin lock here in case the call is made from isr context

Reply:
This driver motivates i2c... Is it even possible to call this from isr? We might need to return an error here.

Reply:
Now it returns `-EWOULDBLOCK` if the read function was called from within an isr.